nifi-nar-bundles/nifi-aws-bundle/nifi-aws-abstract-processors/src/main/java/org/apache/nifi/processors/aws/AbstractAWSProcessor.java

```diff
@@ -425,9 +413,11 @@ protected AWSCredentials getCredentials(final PropertyContext context) {
 
     @OnShutdown
     public void onShutdown() {
-        if ( getClient() != null ) {
+        if (getClient() != null) {
             getClient().shutdown();
         }
+
+        getAwsClientFactory().getClients().forEach(AmazonWebServiceClient::shutdown);
```

Comment:
If we do not store the client separately, only the cached clients need to be shut down.

However, I noticed an old issue with this method: it is hooked on `@OnShutdown`. I believe the intent was `@OnStopped`. Now it runs only at NiFi shutdown which is mostly useless as the JVM will quit and client be destroyed. Also, it shuts down the lastly used client only (clients recreated at start/stop that are not shut down).
As it has not been an issue so far, I think we can simply delete this method for now.
In order to implement the client shutdown properly, we should also handle clients evicted from the cache too. It could be implemented in a separate jira.